app.py

```diff
@@ -45,35 +45,73 @@
 bitbucket_auth = BasicAuth(username=BITBUCKET_USERNAME, password=BITBUCKET_PASSWORD)
-
-# Obtain the access token synchronously
-credentials = {"clientId": PORT_CLIENT_ID, "clientSecret": PORT_CLIENT_SECRET}
-token_response = httpx.post(f"{PORT_API_URL}/auth/access_token", json=credentials)
-port_headers = {"Authorization": f"Bearer {token_response.json()['accessToken']}"}
-
-# Initialize the global AsyncClient with a timeout
 client = httpx.AsyncClient(timeout=httpx.Timeout(60))
 
-async def get_or_create_port_webhook():
-    logger.info("Checking if a Bitbucket webhook is configured on Port...")
+async def get_access_token():
+    credentials = {"clientId": PORT_CLIENT_ID, "clientSecret": PORT_CLIENT_SECRET}
+    token_response = await client.post(f"{PORT_API_URL}/auth/access_token", json=credentials)
+    response_data = token_response.json()
+    access_token = response_data['accessToken']
+    expires_in = response_data['expiresIn']
+    token_expiry_time = datetime.now() + timedelta(seconds=expires_in)
+    return access_token, token_expiry_time
+
+async def refresh_token_if_expired():
+    global port_access_token, token_expiry_time, port_headers
+    if datetime.now() >= token_expiry_time:
+        logger.info("Access token has expired, requesting a new one...")
+        port_access_token, token_expiry_time = await get_access_token()
+        port_headers = {"Authorization": f"Bearer {port_access_token}"}
+        logger.info(f"New token received. Expiry time: {token_expiry_time}")
+
+async def send_port_request(method: str, endpoint: str, payload: Optional[dict] = None):
+    global port_access_token, token_expiry_time, port_headers
+    await refresh_token_if_expired()
+    url = f"{PORT_API_URL}/{endpoint}"
     try:
-        response = await client.get(
-            f"{PORT_API_URL}/webhooks/{WEBHOOK_IDENTIFIER}",
-            headers=port_headers,
-        )
+        response = await client.request(method, url, headers=port_headers, json=payload)
         response.raise_for_status()
-        webhook_url = response.json().get("integration", {}).get("url")
-        logger.info(f"Webhook configuration exists in Port. URL: {webhook_url}")
-        return webhook_url
+        return response
     except httpx.HTTPStatusError as e:
-        if e.response.status_code == 404:
+        if e.response.status_code == 401:
+            # Unauthorized, refresh token and retry
+            logger.info("Received 401 Unauthorized. Refreshing token and retrying...")
+            await refresh_token_if_expired()
+            response = await client.request(method, url, headers=port_headers, json=payload)
+            try:
+                response.raise_for_status()
+                return response
+            except httpx.HTTPStatusError as e:
+                logger.error(f"Error after retrying: {e.response.status_code}")
+                return {"status_code": e.response.status_code, "response": e.response}
```

Comment:
here we will just call refresh_token_and_retry()

if response code == 401
  response = refresh_token_and_retry(args)

Reply:
Done